### PR TITLE
terminology: #[feature] *enables* a feature (instead of "declaring" or "activating" it)

### PR DESCRIPTION
**Pull request by RalfJung**
Mostly, we currently call a feature that has a corresponding `#[feature(name)]` attribute in the current crate a "declared" feature. I think that is confusing as it does not align with what "declaring" usually means. Furthermore, we *also* refer to `#[stable]`/`#[unstable]` as *declaring* a feature (e.g. in [these diagnostics](https://github.com/rust-lang/rust/blob/f25e5abea229a6b6aa77b45e21cb784e785c6040/compiler/rustc_passes/messages.ftl#L297-L301)), which aligns better with what "declaring" usually means. To make things worse, the functions  `tcx.features().active(...)` and  `tcx.features().declared(...)` both exist and they are doing almost the same thing (testing whether a corresponding `#[feature(name)]`  exists) except that `active` would ICE if the feature is not an unstable lang feature. On top of this, the callback when a feature is activated/declared is called `set_enabled`, and many comments also talk about "enabling" a feature.

So really, our terminology is just a mess.

I would suggest we use "declaring a feature" for saying that something is/was guarded by a feature (e.g. `#[stable]`/`#[unstable]`), and "enabling a feature" for  `#[feature(name)]`. This PR implements that.

### CODE REVIEW

File: compiler/rustc_expand/src/config.rs [hunk 1]
```diff
@@ -101,7 +101,7 @@ pub fn features(sess: &Session, krate_attrs: &[Attribute], crate_name: Symbol) -
                 }
             }
 
-            // If the declared feature is unstable, record it.
+            // If the activated feature is unstable, record it.
             if let Some(f) = UNSTABLE_FEATURES.iter().find(|f| name == f.feature.name) {
                 (f.set_enabled)(&mut features);
```
Comment (nnethercote) [inline compiler/rustc_expand/src/config.rs:106]: Should this be called `set_active` now?

**Comment by nnethercote** [comment]
This is an improvement over "declared", but I was going to suggest enable/enabled over activate/active. Several of the existing comments already used enable/enabled.

**Comment by dianne** [comment]
If you're looking to make comments use consistent terminology too, the doc comments in [rustc_middle/src/middle/stability.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/middle/stability.rs) use yet another term: features being "provided".

**Comment by nnethercote** [comment]
It's not a perfect analogy, but [cargo features](https://doc.rust-lang.org/cargo/reference/unstable.html) are "enabled".

**Comment by RalfJung** [comment]
Thinking about it, I think I like "enabled" more.

**Comment by RalfJung** [comment]
I now updated this to "enabling" a feature.

@rustbot ready

File: compiler/rustc_error_codes/src/error_codes/E0636.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
-A `#![feature]` attribute was declared multiple times.
+The same feature is activated multiple times with `#![feature]` attributes
```
Comment (nnethercote) [inline compiler/rustc_error_codes/src/error_codes/E0636.md:1]: This is still using "activated", should be "enabled" in the new world.

File: compiler/rustc_error_codes/src/error_codes/E0636.md [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
 
 Erroneous code example:
 
 ```compile_fail,E0636
 #![allow(stable_features)]
 #![feature(rust1)]
-#![feature(rust1)] // error: the feature `rust1` has already been declared
+#![feature(rust1)] // error: the feature `rust1` has already been activated
```
Comment (nnethercote) [inline compiler/rustc_error_codes/src/error_codes/E0636.md:8]: ditto

File: compiler/rustc_feature/src/unstable.rs [hunk 4]
```diff
@@ -54,61 +54,48 @@ macro_rules! declare_features {
...
-            /// Panics if the symbol doesn't correspond to a declared feature.
-            pub fn active(&self, feature: Symbol) -> bool {
-                match feature {
-                    $( sym::$feature => self.$feature, )*
-
-                    _ => panic!("`{}` was not listed in `declare_features`", feature),
-                }
+            /// Is the given feature enabled (via `#[feature(...)]`)?
+            pub fn enabled(&self, feature: Symbol) -> bool {
+                self.enabled_features.contains(&feature)
```
Comment (nnethercote) [inline compiler/rustc_feature/src/unstable.rs:98]: This doesn't capture features that are enabled-via-edition.

Comment (RalfJung) [reply compiler/rustc_feature/src/unstable.rs:98]: Do we have such features? AFAIK we only have edition-dependent lints, and ofc some hard-coded switches internally, but no direct interaction with the feature machinery. But maybe I missed that.

Anyway I didn't intend to change behavior here. The function should behave just like `declared` did before.

**Comment by nnethercote** [comment]
Oh, I just remembered that I'm responsible for some of the existing naming. It came about in #116437 and #116550, particularly https://github.com/rust-lang/rust/pull/116437/commits/4602d9257de0939ff8770dcf7f10ac1832bce35f and https://github.com/rust-lang/rust/pull/116437/commits/eb209057b1c7a1e7852793e8f53eadfc9d839b2c and https://github.com/rust-lang/rust/pull/116550/commits/d284c8a2d7a2de918a966bee9c9069d7b5bf06bf.

The distinction I used there was the following:
- "declared" features are those for which the program has a `#[feature(...)]` declaration
- "enabled" features are those which are automatically turned on due to the edition
- "active" features = "declared" + "enabled" features

This was better than the previous terminology, which was even more of a mess, but I agree that the distinctions are not obvious and there is room for further improvement.

 "Enabled" seems like the best word for a feature that it turned on and usable by a Rust program. Perhaps "enabled via declaration" vs "enabled via edition" is a better distinction, when that distinction is necessary?

> To make things worse, the functions tcx.features().active(...) and tcx.features().declared(...) both exist and they are doing almost the same thing (testing whether a corresponding #[feature(name)] exists) except that active would ICE if the feature is not an unstable lang feature.

Is that correct? I think this doesn't account for the "enabled via declaration" / "enabled via edition" distinction -- it would fail to succeed for enabled-via-edition features. I'm struggling to remember specific examples of enabled-via-edition features, and how exactly they were enabled. But I think the distinction was important and I am worried that this PR might be eliminating it? Maybe there were some of these features required for the 2018 edition but currently there aren't any? If we are confident that enabled-via-edition features aren't needed any more, then the terminology simplification would be good.

**Comment by RalfJung** [comment]
This is the first time I hear about enabled-by-edition features. Where can I find more about them? The term "edition" does not show up in `unstable.rs`, or in fact anywhere in the rustc_feature crate except in the comments of 2 specific features (and neither indicate that a feature would be "enabled by edition").

Note that all tests still pass, so if there is a distinction here then we have a massive gap in our test suite.

The choice of whether `declared` or `active` is used seemed entirely arbitrary before this PR -- there was not a single comment that I found that would indicate there was ever a conscious choice to select either option.

**Comment by RalfJung** [comment]
The concept of "edition-enabled features" seems to have been removed in https://github.com/rust-lang/rust/pull/118802.

The only place `set_enabled` is called is here

https://github.com/rust-lang/rust/blob/3c374a1a3e62611066fd04a44e8a9870c0fa86c4/compiler/rustc_expand/src/config.rs#L106-L117

and that also calls `set_enabled_lang_feature`.

And the only other place `set_enabled_lang_feature` is called is here

https://github.com/rust-lang/rust/blob/3c374a1a3e62611066fd04a44e8a9870c0fa86c4/compiler/rustc_expand/src/config.rs#L87-L92

and that's for "accepted", i.e. stable, features, so they don't have a `set_enabled`.

But relying on some outside crate to maintain this invariant is not great... also your comments make me wonder whether this is an intentional invariant or an accident.

**Comment by RalfJung** [comment]
I suspect we need those `bool` fields for performance, but let's confirm this.

@bors try
@rust-timer queue

**Comment by nnethercote** [comment]
> The concept of "edition-enabled features" seems to have been removed in https://github.com/rust-lang/rust/pull/118802.

Good to hear. With that known, I agree that "enabled" is a good word choice. I think @ehuss should have a final look over, once the "remove bool" experiment is done:

r? @ehuss

**Comment by RalfJung** [comment]
Oh wow, it basically makes no difference. That's surprising. I guess the hash set is usually empty and so checking it is pretty fast, or so?

I can still bring back the `bool`s as private fields if you prefer, to avoid all these hash set tests.

**Comment by RalfJung** [comment]
@nnethercote ehuss is not on the compiler team, so it seems odd to ask them for review here? This is entirely an internal compiler change.

**Comment by ehuss** [comment]
Yea, I have nothing to say here (other than the perf experiment commit will likely cause a lot of conflicts for other PRs). I'll reassign to the original reviewer.

r? @est31

File: compiler/rustc_query_system/src/ich/impls_syntax.rs [hunk 5]
```diff
@@ -112,10 +112,10 @@ impl<'tcx> HashStable<StableHashingContext<'tcx>> for rustc_feature::Features {
     fn hash_stable(&self, hcx: &mut StableHashingContext<'tcx>, hasher: &mut StableHasher) {
         // Unfortunately we cannot exhaustively list fields here, since the
         // struct is macro generated.
-        self.declared_lang_features.hash_stable(hcx, hasher);
-        self.declared_lib_features.hash_stable(hcx, hasher);
+        self.enabled_lang_features.hash_stable(hcx, hasher);
+        self.enabled_lib_features.hash_stable(hcx, hasher);
 
-        self.all_features()[..].hash_stable(hcx, hasher);
+        self.all_lang_features()[..].hash_stable(hcx, hasher);
```
Comment (nnethercote) [inline compiler/rustc_query_system/src/ich/impls_syntax.rs]: Why the change from `all_features` to `all_lang_features`?

Comment (nnethercote) [reply compiler/rustc_query_system/src/ich/impls_syntax.rs]: Oh, you renamed the function above.

**Comment by nnethercote** [comment]
Ok, r=me for the first two commits. I like the idea of the third one -- good to avoid duplicating the data -- but I think it's worth putting that in a separate PR because it touches so many lines.

**Comment by RalfJung** [comment]
Okay I will move 6c38ef0 to a separate PR.
